storage/engine/mvcc.go

```diff
@@ -1025,6 +1025,204 @@ func MVCCIterate(engine Engine, startKey, endKey proto.Key, timestamp proto.Time
 	return intents, wiErr
 }
 
+// MVCCReverseScan scans the key range specified by end key through start key
+// up to some maximum number of results. Specify max=0 for unbounded
+// scans.
+func MVCCReverseScan(engine Engine, key, endKey proto.Key, max int64, timestamp proto.Timestamp,
+	consistent bool, txn *proto.Transaction, inclusiveKey, inclusiveEndKey bool) ([]proto.KeyValue, []proto.Intent, error) {
+	res := []proto.KeyValue{}
+	intents, err := MVCCReverseIterate(engine, key, endKey, timestamp, consistent, txn, inclusiveKey, inclusiveEndKey,
+		func(kv proto.KeyValue) (bool, error) {
+			res = append(res, kv)
+			if max != 0 && max == int64(len(res)) {
+				return true, nil
+			}
+			return false, nil
+		})
+
+	if err != nil {
+		return nil, nil, err
+	}
+	return res, intents, nil
+}
+
+func isExceedStartKey(metaKey, key proto.EncodedKey, inclusive bool) bool {
+	if inclusive {
+		// If inclusive the start key, the metaKey < key is exceed
+		if bytes.Compare(metaKey, key) < 0 {
+			return true
+		}
+	} else {
+		// If exclusive the start key, the metaKey <= key is exceed
+		if bytes.Compare(metaKey, key) <= 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func isExceedEndKey(metaKey, endKey proto.EncodedKey, inclusive bool) bool {
+	if inclusive {
+		// If inclusive the end key, the metaKey > endKey is exceed
+		if bytes.Compare(metaKey, endKey) > 0 {
+			return true
+		}
+	} else {
+		// If exclusive the end key, the metaKey >= endKey is exceed
+		if bytes.Compare(metaKey, endKey) >= 0 {
+			return true
+		}
+	}
+	return false
+}
+
+// MVCCReverseIterate makes  the iterator backward over the key range specified
+// by start and end keys, At each step of the iteration, f() is invoked with the current
+// key/value pair. If f returns true (done) or an error, the iteration stops and the error
+// is propagated.
+func MVCCReverseIterate(engine Engine, startKey, endKey proto.Key, timestamp proto.Timestamp,
```

Comment:
there is a ton of duplication between this function and MVCCIterate; can you factor out some common code?